docs/transform/routing_transformer.ipynb

```diff
@@ -0,0 +1,457 @@
+    "id": "pIpTv5LXq_9m"
+   },
+   "outputs": [],
+   "source": [
+    "# Add a permutation gate that undoes the action of inserted SWAP gates\n",
+    "initial_qubits, sorted_qubits = zip(*sorted(swap_mapping.items(), key=lambda x: x[1]))\n",
+    "inverse_swap_permutation = [sorted_qubits.index(q) for q in initial_qubits]\n",
+    "routed_circuit.append(cirq.QubitPermutationGate(list(inverse_swap_permutation)).on(*sorted_qubits))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "HHRd-CmqsiQm"
+   },
+   "outputs": [],
+   "source": [
+    "# Reorder to initial physical qubits in `routed_circuit` based on their ordering in `circuit`\n",
+    "_, order = zip(*sorted(list(initial_mapping.items()), key=lambda x: x[0]))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "m_AyaXH8snjG"
+   },
+   "outputs": [],
+   "source": [
+    "print(circuit)\n",
+    "print(routed_circuit)\n",
+    "\n",
+    "# This will not raise an errors now\n",
+    "cirq.testing.assert_allclose_up_to_global_phase(circuit.unitary(), routed_circuit.unitary(qubit_order=order), atol=1e-8)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "RccNfl947CyV"
+   },
+   "source": [
+    "**Note**: the decomposition of the `SwapPermutationGate` is a series of SWAP gates that may be applied on qubits that are not physically adjacent. \n",
+    "\n",
+    "In practice, the user would seldom need to undo the permutation due to inserted SWAP gates as they are often just doing some measurement at the end. Instead, this can be done correctly by just keeping track of the qubit with terminal measurement gate using *measurement keys* (which are unaffected by routing) or by looking at `initial_mapping` and `swap_mapping` to manually trace the permutation of the qubit in question."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "DNmE3napAXzg"
+   },
+   "source": [
+    "## Extending the Routing API"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "mvLpskslAdEa"
+   },
+   "source": [
+    "### By Overriding the Heuristic Cost Function"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "tV6Ms1CVtpJv"
+   },
+   "source": [
+    "A user may decide that the existing cost function in `RouteCQC` can be improved or they may have a cost function that performs better in particular cases (injecting noise awareness, working devices with fixed topologies, etc.) We provide an easy to override the existing cost function by means of class extension. For example,"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "yqWstLnSuSw6"
+   },
+   "outputs": [],
+   "source": [
+    "from typing import Tuple, List, Any\n",
+    "\n",
+    "QidPair = QidPair = Tuple['cirq.Qid', 'cirq.Qid']\n",
+    "\n",
+    "class RouteCQCSimpleCostFunction(cirq.RouteCQC):\n",
```

Comment:
It would be nice if we can show that this derived class actually works (so use an instance of this derived class to route a circuit and print it. We don't need to verify it etc.